src/v/cloud_topics/core/event_filter.h

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "cloud_topics/core/pipeline_stage.h"
+#include "container/intrusive_list_helpers.h"
+
+#include <seastar/core/expiring_fifo.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/lowres_clock.hh>
+
+namespace experimental::cloud_topics::core {
+
+enum class event_type {
+    // This indicates default initialized event
+    none,
+    // New read request registered
+    read_request,
+    // New write request registered
+    write_request,
+    // Pipeline is shutting down
+    shutting_down,
+    // Filter timed out
+    timeout,
+};
+
+struct event {
+    pipeline_stage stage;
+    event_type type{event_type::none};
+    size_t pending_write_bytes{0};
+    size_t total_write_bytes{0};
+    size_t pending_read_bytes{0};
+    size_t total_read_bytes{0};
+};
+
+/// This is similar to remote::event_filter in
+/// the cloud_storage. It's used to trigger different activities
+/// by events.
+/// E.g. the background upload loop in the batcher can start L0
+/// upload early if enough data is accumulated.
+template<class Clock = ss::lowres_clock>
```

Comment:
I wonder if it is possible to templatize like for example `event_filter<event_type::write_request>`? Don't have enough metaprogramming experience to asses whether this is possible without digging deeper.

this way when you co_await c.subscribe(...) you won't need a switch or when you subscribe with timeout you would need to switch but only on write_request and timeout rather than on write requests, shutdown (this can also be just a timeout event) so in the normal case you could have 2 events.

Or just propagate timeout, shutdown as exception and have no switch at all.